src/osd/OSDMap.cc

```diff
@@ -1622,6 +1694,40 @@ int OSDMap::_pick_primary(const vector<int>& osds) const
   return -1;
 }
 
+void OSDMap::_apply_remap(const pg_pool_t& pi, pg_t raw_pg, vector<int> *raw) const
+{
+  pg_t pg = pi.raw_pg_to_pg(raw_pg);
+  auto p = pg_remap.find(pg);
+  if (p != pg_remap.end()) {
+    *raw = p->second;
+    return;
+  }
+
+  auto q = pg_remap_items.find(pg);
+  if (q != pg_remap_items.end()) {
+    // NOTE: this approach does not allow a bidirectional swap,
```

Comment:
it also wouldn't handle a degenerate ec case, where the same osd appears more than once - e.g. [1, 0, 2, -1, 1] - or can that only happen in the acting remappings?